core/src/main/java/io/methvin/watcher/DirectoryWatcher.java

```diff
@@ -224,15 +242,28 @@ public DirectoryWatcher(
    * is closed.
    */
   public void watch() {
+    final AtomicInteger eventCount = new AtomicInteger(0);
     while (listener.isWatching()) {
       // wait for key to be signalled
       WatchKey key;
       try {
-        key = watchService.take();
+        key = watchService.poll();
+        if (key == null) {
+          if (timeout > 0) {
+            currentOnIdleTimerTask = new OnIdleTimerTask(listener, eventCount.get());
+            timer.schedule(currentOnIdleTimerTask, timeout);
+          }
+          key = watchService.take();
+        }
       } catch (InterruptedException x) {
         return;
       }
       for (WatchEvent<?> event : key.pollEvents()) {
+        eventCount.incrementAndGet();
+        if (currentOnIdleTimerTask != null && !currentOnIdleTimerTask.isDone) {
```

Comment:
Does this need to be in the 'for' loop? Isn't it enough to do this  directly after the take() returns?

I don't think this is thread safe, as the timer will set isDone, which is different to the directory watcher main loop thread.